### PR TITLE
Generic crawler

### PR DESCRIPTION
**Pull request by ammongit**
This modifies the crawler to instead be an ABC that can be extended. This will make implementing the audit log down the line easier and more modular.

### CODE REVIEW

File: statbot/sql.py [hunk 1]
```diff
@@ -28,6 +28,8 @@
 
 Column = functools.partial(Column, nullable=False)
 
+MAX_ID = (1 << 63) - 1
```
Comment (raylu) [inline statbot/sql.py]: can we just write `2**64 - 1`?

File: statbot/crawler.py [hunk 2]
```diff
@@ -80,25 +80,30 @@ def start(self):
         yield_delay = self.config['crawler']['yield-delay']
         long_delay = self.config['crawler']['empty-source-delay']
 
+        done = {source: False for source in self.progress.keys()}
```
Comment (raylu) [inline statbot/crawler.py]:
```python
done = dict.fromkeys(self.progress.keys(), False)
```

File: statbot/crawler.py [hunk 3]
```diff
@@ -11,165 +11,201 @@
...
+                        await self.queue.put((source, None, NOW_ID))
+                        self.progress[source] = NOW_ID
+                    else:
+                        # This source still has more
+                        last_id = self.get_last_id(events)
+                        await self.queue.put((source, events, last_id))
+                        self.progress[source] = last_id
                 except Exception:
-                    self.logger.error(f"Error reading (or syncing) messages from channel id {cid}", exc_info=1)
+                    self.logger.error(f"Error reading events from source {source}", exc_info=1)
```
Comment (raylu) [inline statbot/crawler.py]: this could be just `except`

don't you want to completely bail from this source rather than get right back in the loop?

Comment (ammongit) [reply statbot/crawler.py]: Well the problem might not be with the source. Perhaps one particular message has some unusual property I hadn't thought about that prevents it from being inserted, but otherwise the crawling is totally fine.

Comment (raylu) [reply statbot/crawler.py]: that is possible, but we're catching a very broad exception. let me rephrase:

if a bug is preventing us from crawling a source, doesn't that ruin everything and shouldn't we fix that by bailing from this loop? alternatively, given that we don't even know why this might fail, instead of catching the exception shouldn't we just bail from the program entirely?

Comment (ammongit) [reply statbot/crawler.py]: Yes, I understand why we might want to have it fail on exceptions, but the possibility of otherwise innocuous SQL issues is possible.

To try and strike a balance here, I'm thinking we could have it catch and log SQLAlchemy exceptions, but fail on everything else.

Comment (ammongit) [reply statbot/crawler.py]: Alright I have changed this. Now only the consumer can catch exceptions, and when it does, only SQL-related things are caught. Any other kind of exception will terminate the bot.

File: statbot/crawler.py [hunk 4]
```diff
@@ -11,165 +11,200 @@
...
+            source, events, last_id = await self.queue.get()
+            self.logger.info(f"{self.name}: got group of events from queue")
+
+            try:
+                with self.sql.transaction() as trans:
+                    if events is not None:
+                        await self.write(trans, events)
+                    await self.update(trans, source, last_id)
+            except SQLAlchemyError:
+                self.logger.error(f"{self.name}: error during event write", exc_info=1)
```
Comment (raylu) [inline statbot/crawler.py]: I'm still not a huge fan of this since we don't have a specific idea of the conditions that caused this and whether or not it's safe to continue, nor will we likely notice because the bot will just keep running, chundering logs into the ether, but it's better than a bare except I suppose
